src/plugins/controls/public/control_group/control_group_renderer.tsx

```diff
@@ -8,41 +8,87 @@
-import React, { useEffect, useMemo, useRef, useState } from 'react';
+import React, { useMemo, useRef, useState } from 'react';
 
 import { IEmbeddable } from '@kbn/embeddable-plugin/public';
+import { useReduxContainerContext } from '@kbn/presentation-util-plugin/public';
 
 import { pluginServices } from '../services';
-import { getDefaultControlGroupInput } from '../../common';
-import { ControlGroupInput, ControlGroupOutput, CONTROL_GROUP_TYPE } from './types';
+import { ControlPanelState, getDefaultControlGroupInput } from '../../common';
+import {
+  ControlGroupInput,
+  ControlGroupOutput,
+  ControlGroupReduxState,
+  CONTROL_GROUP_TYPE,
+} from './types';
 import { ControlGroupContainer } from './embeddable/control_group_container';
+import { DataControlInput } from '../types';
+import { getCompatibleControlType, getNextPanelOrder } from './embeddable/control_group_helpers';
+import { controlGroupReducers } from './state/control_group_reducers';
+
+const ControlGroupInputBuilder = {
+  addDataControlFromField: async (
+    initialInput: Partial<ControlGroupInput>,
+    newPanelInput: {
+      title?: string;
+      panelId?: string;
+      fieldName: string;
+      dataViewId: string;
+    } & Partial<ControlPanelState>
+  ) => {
+    const { defaultControlGrow, defaultControlWidth } = getDefaultControlGroupInput();
+    const controlGrow = initialInput.defaultControlGrow ?? defaultControlGrow;
+    const controlWidth = initialInput.defaultControlWidth ?? defaultControlWidth;
+
+    const { panelId, dataViewId, fieldName, title, grow, width } = newPanelInput;
+    const newPanelId = panelId || uuid.v4();
+    const nextOrder = getNextPanelOrder(initialInput);
+    const controlType = await getCompatibleControlType({ dataViewId, fieldName });
+
+    initialInput.panels = {
+      ...initialInput.panels,
+      [newPanelId]: {
+        order: nextOrder,
+        type: controlType,
+        grow: grow ?? controlGrow,
+        width: width ?? controlWidth,
+        explicitInput: { id: newPanelId, dataViewId, fieldName, title: title ?? fieldName },
+      } as ControlPanelState<DataControlInput>,
+    };
+  },
+};
 
 export interface ControlGroupRendererProps {
-  input?: Partial<Pick<ControlGroupInput, 'viewMode' | 'executionContext'>>;
   onEmbeddableLoad: (controlGroupContainer: ControlGroupContainer) => void;
+  getCreationOptions: (
+    builder: typeof ControlGroupInputBuilder
+  ) => Promise<Partial<ControlGroupInput>>;
 }
 
-export const ControlGroupRenderer = ({ input, onEmbeddableLoad }: ControlGroupRendererProps) => {
+export const ControlGroupRenderer = ({
+  onEmbeddableLoad,
+  getCreationOptions,
+}: ControlGroupRendererProps) => {
   const controlsRoot = useRef(null);
   const [controlGroupContainer, setControlGroupContainer] = useState<ControlGroupContainer>();
-
   const id = useMemo(() => uuid.v4(), []);
-
   /**
    * Use Lifecycles to load initial control group container
    */
   useLifecycles(
```

Comment:
I think that while I was working on Portable Dashboard I ran into some problem with the `useLifecycles` hook which made me replace it with a general `useEffect`. Not sure exactly what the problem was anymore because it was a while ago, but I'd make sure that it gets called when you expect it to, and that the cleanup function works as / when expected.